pkg/ast/spl/spl.peg

```diff
@@ -586,6 +586,81 @@ TextExpr <- (opName:("lower") "(" stringExpr:StringExpr ")") {
     }
     return node, nil
 }
+/ (opName:("substr") "(" stringExpr:StringExpr ", " startIndex:NumericExpr lengthParam:(", " NumericExpr)? ")") {
+    opNameStr, err := transferUint8ToString(opName)
+    if err != nil {
+        return nil, fmt.Errorf("Spl peg: TextExpr: %v", err)
+    }
+    stringExprConverted, ok := stringExpr.(*structs.StringExpr)
+    if !ok {
+        return nil, fmt.Errorf("Spl peg: TextExpr: Failed to assert stringExpr as *structs.StringExpr")
+    }
+    startIndexConverted, ok := startIndex.(*structs.NumericExpr)
+    if !ok {
+        return nil, fmt.Errorf("Spl peg: TextExpr: Failed to assert startIndex as *structs.NumericExpr")
+    }
+  var lengthConverted *structs.NumericExpr
+    if lengthParam != nil {
+        lengthSlice, ok := lengthParam.([]interface{})
+        if ok && len(lengthSlice) > 1 {
+            length, ok := lengthSlice[1].(*structs.NumericExpr)
+            if ok {
+                lengthConverted = length
+            } else {
+                return nil, fmt.Errorf("Spl peg: TextExpr: Unable to assert length as *structs.NumericExpr")
+            }
+        }
+    }
+    node := &structs.TextExpr {
+        Op: opNameStr,
+        Value: stringExprConverted,
+        StartIndex: startIndexConverted,
+        LengthExpr: lengthConverted,
+        
+    }
+    return node, nil
+}
+/ (opName:("tonumber") "(" stringExpr:StringExpr baseExpr:("," NumericExpr)? ")") {
+    opNameStr, err := transferUint8ToString(opName)
+    if err != nil {
+        return nil, fmt.Errorf("Spl peg: TextExpr: %v", err)
+    }
+    stringExprConverted, ok := stringExpr.(*structs.StringExpr)
+    if !ok {
+        return nil, fmt.Errorf("Spl peg: TextExpr: Failed to assert stringExpr as *structs.StringExpr")
+    }
+    var baseExprConverted *structs.NumericExpr
+    if baseExpr != nil {
+        baseSlice, ok := baseExpr.([]interface{})
+        if ok && len(baseSlice) > 1 {
+            base, ok := baseSlice[1].(*structs.NumericExpr)
+            if ok {
+                baseExprConverted = base
+            } else {
+                return nil, fmt.Errorf("Spl peg: TextExpr: Unable to assert base as *structs.NumericExpr")
+            }
+        }
+    } 
+    node := &structs.TextExpr {
+        Op: opNameStr,
+        Value: stringExprConverted,
+        BaseExpr: baseExprConverted,
+    }
+    return node, nil
+}
```

Comment:
Shouldn't this be in the `NumericExpr` rule rather than the `TextExpr` rule? I think the `TextExpr` is only for things that return strings and the `NumericExpr` is for things that return numbers.

Reply:
As per the documentation tonumber takes a string value hence wrote it under text expr
https://docs.splunk.com/Documentation/SCS/current/SearchReference/ConversionFunctions#tonumber.28.26lt.3Bstr.26gt.3B.2C_.26lt.3Bbase.26gt.3B.29

Reply:
Ok, but I think we want to write rules based on the value they return. That way, when a rule has a `TextExpr` as a subpart of the rule, we know the `TextExpr` part will return something with a string value. The `tonumber()` function should return a `NumericExpr` struct, but the `TextExpr` rule should match things that return strings. So I think this `tonumber()` parsing should be moved to the `NumericEvalExpr` rule